flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartition.java

```diff
@@ -149,7 +149,11 @@ private void writeAndCloseBufferConsumer(BufferConsumer bufferConsumer) throws I
 		try {
 			final Buffer buffer = bufferConsumer.build();
 			try {
-				data.writeBuffer(buffer);
+				if (bufferCompressor != null && buffer.isBuffer() && buffer.readableBytes() > 0) {
+					data.writeBuffer(bufferCompressor.compressToInternalBuffer(buffer));
```

Comment:
As mentioned above, this can be rewritten as 
```
if (canBeCompressed(buffer)) {
    parent.bufferCompressor. compressToInternalBuffer(buffer);
}
```